### PR TITLE
Accumulating MACs after reshare + test

### PR DESCRIPTION
**Pull request by richajaindce**
1. Accumulating MACs after reshare
2. Reshare malicious test which fails validation check on additive attack

### CODE REVIEW

File: src/test_fixture/world.rs [hunk 1]
```diff
@@ -230,7 +230,8 @@ where
 
         // Perform validation and convert the results we just got: P to O
         let output = join_all(
-            zip(v, m_results).map(|(v, m_result)| async { v.validate(m_result).await.unwrap() }),
+            zip(v, m_results)
+                .map(|(v, m_result)| async move { v.validate(m_result).await.unwrap() }),
```
Comment (benjaminsavage) [inline src/test_fixture/world.rs:234]: Why did you need to modify this line? Why add this "move"?

File: src/protocol/malicious.rs [hunk 2]
```diff
@@ -197,7 +197,7 @@ impl<'a, F: Field> MaliciousValidator<'a, F> {
     ///
     /// ## Panics
     /// Will panic if the mutex is poisoned
-    pub async fn validate<D: DowngradeMalicious>(self, values: D) -> Result<D::Target, Error> {
+    pub async fn validate<D: DowngradeMalicious>(&self, values: D) -> Result<D::Target, Error> {
```
Comment (benjaminsavage) [inline src/protocol/malicious.rs:200]: @martinthomson says we should not do this.

File: src/protocol/sort/reshare.rs [hunk 3]
```diff
@@ -104,7 +108,7 @@ impl<F: Field> Reshare<F> for MaliciousContext<'_, F> {
         use crate::secret_sharing::ThisCodeIsAuthorizedToDowngradeFromMalicious;
         let rx_ctx = self.narrow(&ReshareMAC);
         let (x, rx) = try_join(
-            self.semi_honest_context().reshare(
+            self.narrow(&ReshareInput).semi_honest_context().reshare(
```
Comment (benjaminsavage) [inline src/protocol/sort/reshare.rs:111]: Why is this needed? I don't think this is necessary. The caller should have narrowed (if necessary) before invoking "reshare"

File: src/protocol/sort/reshare.rs [hunk 4]
```diff
@@ -213,5 +225,69 @@ mod tests {
...
+            input: &MaliciousReplicated<F>,
+            record_id: RecordId,
+            to_helper: Role,
+            additive_error: F,
+        ) -> Result<MaliciousReplicated<F>, Error> {
+            use crate::protocol::context::SpecialAccessToMaliciousContext;
+            use crate::secret_sharing::ThisCodeIsAuthorizedToDowngradeFromMalicious;
+            let rx_ctx = ctx.narrow("reshare_mac");
+            let (x, rx) = try_join(
+                ctx.narrow("reshare_input").semi_honest_context().reshare(
```
Comment (benjaminsavage) [inline src/protocol/sort/reshare.rs:240]: I don't think you need to narrow here.

File: src/protocol/sort/reshare.rs [hunk 5]
```diff
@@ -213,5 +225,69 @@ mod tests {
...
+        async fn reshare_with_additive_attack<F: Field>(
+            ctx: MaliciousContext<'_, F>,
+            input: &MaliciousReplicated<F>,
+            record_id: RecordId,
+            to_helper: Role,
+            additive_error: F,
+        ) -> Result<MaliciousReplicated<F>, Error> {
+            use crate::protocol::context::SpecialAccessToMaliciousContext;
+            use crate::secret_sharing::ThisCodeIsAuthorizedToDowngradeFromMalicious;
+            let rx_ctx = ctx.narrow("reshare_mac");
```
Comment (benjaminsavage) [inline src/protocol/sort/reshare.rs:238]: I think it would be better to use `&Step::ReshareMAC`

File: src/protocol/sort/reshare.rs [hunk 6]
```diff
@@ -213,5 +225,69 @@ mod tests {
...
+                    record_id,
+                    to_helper,
+                ),
+                rx_ctx
+                    .semi_honest_context()
+                    .reshare(input.rx(), record_id, to_helper),
+            )
+            .await?;
+            let malicious_input =
+                MaliciousReplicated::new(x + &Replicated::new(additive_error, additive_error), rx);
```
Comment (benjaminsavage) [inline src/protocol/sort/reshare.rs:251]: So this is not actually the place where the additive attack happens - that's on the communication with the other helpers - where it sends a value.

**Review by benjaminsavage — CHANGES_REQUESTED** [review]

File: src/protocol/sort/reshare.rs (hunk 3 above)
Comment (richajaindce) [reply src/protocol/sort/reshare.rs:111]: We need to reshare both x and rx at this step. We also need context to calculate accumulate_macs. self.semi_honest will consume self. Another option is to clone self before resharing x.

File: src/protocol/sort/reshare.rs (hunk 6 above)
Comment (richajaindce) [reply src/protocol/sort/reshare.rs:251]: U are right. Let me change this

File: src/test_fixture/world.rs (hunk 1 above)
Comment (richajaindce) [reply src/test_fixture/world.rs:234]: If got left from changes. Apologies

File: src/protocol/sort/reshare.rs [hunk 7]
```diff
@@ -213,5 +224,120 @@ mod tests {
...
+                    .await?;
+
+                // Sleep until `to_helper.left` sends us their part1 value
+                let part1: F = channel
+                    .receive(to_helper.peer(Direction::Left), record_id)
+                    .await?;
+
+                Ok(Replicated::new(r0, part1 + part2 + additive_error))
+            } else {
+                Ok(Replicated::new(r0 + additive_error, r1 - additive_error))
```
Comment (benjaminsavage) [inline src/protocol/sort/reshare.rs]: This helper cannot wage an additive attack

**Review by benjaminsavage — APPROVED** [review]
I love it!
